sdl_viewer/src/graphic/moving_window_texture.rs

```diff
@@ -0,0 +1,364 @@
+    ) -> [UpdateRegion<'a, P>; 4] {
+        // Do subtractions in signed numbers to avoid overflow
+        let width: i32 = pixels.width().try_into().unwrap();
+        let height: i32 = pixels.height().try_into().unwrap();
+        let width_1_3 = width.min(size - xoff);
+        let width_2_4 = width - width_1_3;
+        let height_1_2 = height.min(size - yoff);
+        let height_3_4 = height - height_1_2;
+        // The image crate likes unsigned ints
+        let width_1_3 = u32::try_from(width_1_3).unwrap();
+        let width_2_4 = u32::try_from(width_2_4).unwrap();
+        let height_3_4 = u32::try_from(height_3_4).unwrap();
+        let height_1_2 = u32::try_from(height_1_2).unwrap();
+        [
+            UpdateRegion {
+                x: xoff,
+                y: yoff,
+                pixels: pixels.view(0, 0, width_1_3, height_1_2),
+            },
+            UpdateRegion {
+                x: 0,
+                y: yoff,
+                pixels: pixels.view(width_1_3, 0, width_2_4, height_1_2),
+            },
+            UpdateRegion {
+                x: xoff,
+                y: 0,
+                pixels: pixels.view(0, height_1_2, width_1_3, height_3_4),
+            },
+            UpdateRegion {
+                x: 0,
+                y: 0,
+                pixels: pixels.view(width_1_3, height_1_2, width_2_4, height_3_4),
+            },
+        ]
+    }
+}
+
+impl<P> GlMovingWindowTexture<P>
+where
+    P: TextureFormat,
+    P: Pixel + 'static + std::fmt::Debug,
+    P::Subpixel: 'static + std::fmt::Debug,
+{
+    pub fn new(
+        program: &GlProgram,
+        gl: Rc<opengl::Gl>,
+        name: &str,
+        size: i32,
+        pixels: ImageBuffer<P, Vec<P::Subpixel>>,
+    ) -> Self {
+        let mut id = 0;
+        unsafe {
+            gl.UseProgram(program.id);
+            let loc =
+                gl.GetUniformLocation(program.id, (name.to_string() + "\0").as_ptr() as *const i8);
+            gl.Uniform1i(loc, 0);
+
+            gl.GenTextures(1, &mut id);
+            gl.BindTexture(opengl::TEXTURE_2D, id);
+
+            // Enable wraparound addressing
+            gl.TexParameteri(
+                opengl::TEXTURE_2D,
+                opengl::TEXTURE_WRAP_S,
+                opengl::REPEAT as i32,
+            );
+            gl.TexParameteri(
+                opengl::TEXTURE_2D,
+                opengl::TEXTURE_WRAP_T,
+                opengl::REPEAT as i32,
+            );
+
+            // Intended for exact pixel fetching, no interpolation needed
+            gl.TexParameteri(
+                opengl::TEXTURE_2D,
+                opengl::TEXTURE_MIN_FILTER,
+                opengl::NEAREST as i32,
+            );
+            gl.TexParameteri(
+                opengl::TEXTURE_2D,
+                opengl::TEXTURE_MAG_FILTER,
+                opengl::NEAREST as i32,
+            );
+            gl.TexImage2D(
+                opengl::TEXTURE_2D,
+                0,
```

Comment:
nit: add short comments what these `0` arguments are? Alternatively you could define new constants, for example `BASE_IMAGE_LEVEL = 0` and use them instead of the magic number.

Reply:
I added a comment like above